### PR TITLE
[Raft] Add a transaction janitor

### PR DESCRIPTION
**Pull request by luluz66**
**Related issues**: 
https://github.com/buildbuddy-io/buildbuddy-internal/issues/3279

- Add a proto `TxnRecord` and store it in meta range at different stages of
`RunTxn`. 
- Add a transaction janitor process to read the txn record and help get the
  stuck transactions unstuck.

### CODE REVIEW

File: enterprise/server/raft/constants/constants.go [hunk 1]
```diff
@@ -70,6 +75,9 @@ var (
 	// The last rangeID that was generated.
 	LastRangeIDKey = keys.MakeKey(SystemPrefix, []byte("last_range_id"))
 
+	// A prefix to prepend to transaction records.
+	TxnRecordPrefix = keys.MakeKey(SystemPrefix, []byte("txn-record-"))
```
Comment (tylerwilliams) [inline enterprise/server/raft/constants/constants.go:79]: nit: is it worth specifying that it's a coordinator record to distinguish from replica written value?

File: enterprise/server/raft/rbuilder/rbuilder.go [hunk 2]
```diff
@@ -51,6 +51,10 @@ func (bb *BatchBuilder) Add(m proto.Message) *BatchBuilder {
 		req.Value = &rfpb.RequestUnion_DirectWrite{
 			DirectWrite: value,
 		}
+	case *rfpb.DirectDeleteRequest:
+		req.Value = &rfpb.RequestUnion_DirectDelete{
+			DirectDelete: value,
+		}
```
Comment (tylerwilliams) [inline enterprise/server/raft/rbuilder/rbuilder.go:57]: can you add this in a separate CL?

File: enterprise/server/raft/txnjanitor/txnjanitor.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	txnLivessnessThreshold = 10 * time.Second
+	txnCleanupPeriod       = 15 * time.Second
+)
+
+type IStore interface {
+	IsLeader(shardID uint64) bool
+	Sender() *sender.Sender
+}
+
+type TxnJanitor struct {
```
Comment (tylerwilliams) [inline enterprise/server/raft/txnjanitor/txnjanitor.go:28]: we should add tests for this functionality

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
overall this looks good -- just a few comments!

**Review by tylerwilliams — COMMENTED** [review]
it's kind of messy how DeleteTxnRecord and WriteTxnRecord are in the sender but called from the janitor. Could all of these txn coordinator methods be moved to a txncoordinator file, along with the janitor code?

**Comment by luluz66** [comment]
> it's kind of messy how DeleteTxnRecord and WriteTxnRecord are in the sender but called from the janitor. Could all of these txn coordinator methods be moved to a txncoordinator file, along with the janitor code?
I moved the txn related methods including RunTxn from sender to txn.go.

File: enterprise/server/raft/txn/txn.go [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+	// How often do we scan transaction records and clean them up.
+	txnCleanupPeriod = 15 * time.Second
+)
+
+type IStore interface {
+	IsLeader(shardID uint64) bool
+	Sender() *sender.Sender
+}
+
+type TxnCoordinator struct {
```
Comment (tylerwilliams) [inline enterprise/server/raft/txn/txn.go:37]: nit: just call this Coordinator

**Review by tylerwilliams — APPROVED** [review]

**Pull request merged**